src/Repositories/MySql/MySql.php

```diff
@@ -282,6 +283,41 @@ public function batchCommitUpdatesFromCollection(RepositoryCollection $collectio
         static::executeStatement($sql, $namedParams);
     }
 
+    /**
+     * Provides an opportunity for a collection's size to be calculated in an optimised way
+     * 
+     * By default just gets a collection cursor and counts the rows.
+     */
+    public function countRowsInCollection(RepositoryCollection $collection)
+    {        
+        $clone = clone $collection;
+        $clone->disableRanging();
+
+        $sqlStatement = $this->getSqlStatementForCollection($clone, $params);        
+        
+        if ($sqlStatement->columns[0] instanceof SelectExpression){
+            if (stripos($sqlStatement->columns[0]->expression, "*") !== false) {
+                // The query includes a wild card select on the outermost table. In some instances there can
+                // be a pull up that clashes with column names in the outermost table. This is okay for a normal
+                // result set (right most column wins) but when you wrap with a SELECT COUNT(*) MySql complains about
+                // the duplicate column. Easiest solution is to change the * for just the ID.
+                //
+                // Note that while we don't care about the data in all pull up columns we can't remove them as sometimes
+                // orders and having clauses expect them to be in the select list.
+                //
+                // Note also that we alias the ID as uniqid() as sometimes the pull ups are pulling up an identifier column
+                // from a joined table and the names can still clash.
+                $sqlStatement->columns[0]->expression = str_replace("*", $collection->getModelSchema()->uniqueIdentifierColumnName." AS ".uniqid(), $sqlStatement->columns[0]->expression);                                
+            }
+        }
+
+        $sqlStatement = "SELECT COUNT(*) AS rowCount FROM (".$sqlStatement.") AS countableList";
+        $connection = static::getReadOnlyConnection();        
+        $statement = static::executeStatement((string)$sqlStatement, $params, $connection);
+        $row = $statement->fetch(PDO::FETCH_ASSOC);
+        return $row["rowCount"];
+    }
+
```

Comment:
I think we might be able to use `$collection->count()` in `createCursorForCollection` now instead of both `SQL_CALC_FOUND_ROWS` and `SELECT FOUND_ROWS()` where `$hasLimit` as those are now (i hope) redundant 🤞 ?

I gave it a go locally and nothing caught fire 🔥  🚒  🧑‍🚒  🧯 . I take that as a good sign.

Reply:
The only benefit of keeping the calc found rows and found_rows() is that (in a commit made this morning) now if you have accessed the collection data first and later do a count, we can avoid a second query as the result set is already loaded along with the full row size. I'm happy to leave this in for now as it's not causing any perceived issues and hopefully avoids secondary cases in a small number of cases.

Reply:
Sorry, i should've been clearer about the issue - if (on a collection) i do a count, before an iteration, the code runs the new count query. then it (i think unnecessarily) runs the `SELECT FOUND_ROWS()` query when it comes to create the cursor.

You're quite right though, if you iterate then count, I believe found rows will be a more efficient way to get the count in that case.

I think the issue may be that the cursor doesn't check to see if count has been done yet on the collection. Could it just do something to check `$collection->count !== null` to determine that, and just not re-do the count if the count has been done by `countRowsInCollection`?

Reply:
Thanks Sam - yes we could look to optimise for if a count() comes first and we aren't using a range. For now however I'm going to merge this as Linergy have found no issues and the performance gain was substantial.